### ISSUE TITLE
Add jQuery 4.x as additional option

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#6404]
The first 4.x Beta for jQuery has been [published recently](https://blog.jquery.com/2024/02/06/jquery-4-0-0-beta/).

It brings several changes that affect core code, which isn't fully compatible, yet. We should start to prepare for a stable 4.x release.

**Are we there, yet?** Short answer: no.

We have some (not many) usages of deprecated (and removed in 4) functions, covered in https://github.com/backdrop/backdrop-issues/issues/6248 :point_left: I'd suggest to start there.

We have to decide on a strategy with **third party code** that isn't compatible anymore and isn't maintained anymore:

- jquery.form.js
- jquery.cookie.js
- jquery.timeentry.js (see also #6249)
- ???

While we might consider to just drop jquery.cookie.js and jquery.timeentry.js, that's no option for jquery.form.js.
Note that Drupal struggles with a similar problem: https://www.drupal.org/project/drupal/issues/3419734 Let's keep an eye on their considerations.

**Comment by indigoxela** [comment]
Another idea: fix all problems in our own code, then check which removed jQuery functions are missing re 3rd party stuff and provide those functions as bc-layer in core.

I know, that's just postponing the problem, but... that would allow us to integrate jQuery v4 faster without having to fork the unmaintained stuff, or even decide whether we should fork.

**Comment by indigoxela** [comment]
A [PR](https://github.com/backdrop/backdrop/pull/4657) is available to play with.

To see jQuery 4 in action, go to admin/config/development/jquery, check "jQuery 4 (experimental)" and save.

On that same page you can open a browser console and run `jQuery.fn.jquery`. That will output "4.0.0-beta".

Click around, do some AJAX stuff (with nodes, fields, layouts, views...). In theory everything works.
Nothing's decided, yet, and there are many open questions.

- Do we want such a bc layer for plugins?
- More of it? (Bring back more of those removed utility functions?)
- Do we want yet another jQuery version to choose from on admin/config/development/jquery?
- What do we do with v5, v6...?
- When do we want to make v4 available in Backdrop?
- ...

As jQuery 4 is still Beta, anyway, we have some time to make up our minds.

**Comment by klonos** [comment]
Thank you @indigoxela 🙏🏼 ❤️ 

I don't have all the answers to the above questions, but here are some thoughts:

> Do we want yet another jQuery version to choose from on admin/config/development/jquery?

I can't think any reason why that would be a problem. That form is not "busy", and besides, people would only visit that once every so to configure the version they want, then forget about it if their site works fine.

> What do we do with v5, v6...?

We should add telemetry on this, then as new versions of jQuery are coming out and older ones get deprecated/EoL'ed we can consider removing/adding versions as needed. We can change the default for new installations as soon as one version has been tested enough and any bugs have been surfaced/squashed. And finally, we can consider moving old versions to contrib.

> When do we want to make v4 available in Backdrop?

As soon as there's a stable release I think. We can still keep it with the "experimental" indication, and leave v3 as the default. Then as we get more confident, we can change the default to v4 in a future core release. How confidence will be gained:
- telemetry
- perhaps consider switching our PR sandboxes to be using v4 by default (using https://github.com/backdrop/backdrop/blob/1.x/.github/misc/settings.local.php to override things), which will give us more exposure to testing and discovering any possible bugs (and if v4 is not working as expected with a specific PR, then it'd be easy to switch to v3 and carry on).

> ...we have some time to make up our minds.

Yes 👍🏼 ...and even if v4 stable gets released, we don't have to rush then either.

**Comment by indigoxela** [comment]
@klonos many thanks for your thoughts, that's mostly, what I had in mind. :pray: 

Another thing to keep in mind: contrib. Many modules aren't even fully compatible with 3.x, yet. So I think, there's still a way to go, before we could consider switching the default (for new sites) to 4.x. But as you stated - no need to rush.

**Comment by indigoxela** [comment]
Something else to consider re my proposal to add a (small) bc layer for jQuery 4: contrib projects.

I'd assume, that many contrib modules aren't fully compatible with jQuery 3, yet, let alone jQuery 4.
By providing that layer, we'd enable these projects to be used with v4 without breaking them. However, currently the bc layer contains only the required functions for (3rd party) plugins used in core, not all removed functions.

**Comment by indigoxela** [comment]
Time for a rebase - done. :wink: 

Currently the [jQuery 4.0.0 milestone](https://github.com/jquery/jquery/milestone/7) has only one open issue. But, of course, there's no due date. But let's keep an eye on it.

**Comment by indigoxela** [comment]
Oh, jQuery.ui got updated: https://blog.jqueryui.com/2024/05/jquery-ui-1-14-0-beta-1-released/
Seems like it was also tested against the jQuery 4.x dev branch, which is good news. Although I think, it was compatible before, this makes sure, jQuery.UI plays nicely with v4.

**Comment by indigoxela** [comment]
Rebased to get back a PR sandbox - for a very short time. :wink:

**Comment by indigoxela** [comment]
They just announced the release of a [Second Beta of jQuery 4.0.0](https://blog.jquery.com/2024/07/17/second-beta-of-jquery-4-0-0/).

Most of the changes seem related test and build infrastructure: https://github.com/jquery/jquery/releases/tag/4.0.0-beta.2

On Mastodon it seems like they're leaning towards "RC next"? Maybe... https://social.lfx.dev/@jquery/112803865842862230 :wink: 

So I updated my experimental PR - it needed another rebase, anyway. Happy testing!

**Comment by indigoxela** [comment]
It was time for another rebase, fresh sandbox available.

It came as a surprise to me, that D11 already ships with jQuery 4.0.0-beta, but that's off topic here. 😉

**Comment by argiepiano** [comment]
Thanks @indigoxella. Is there a list of items for manual testing somewhere?

**Comment by indigoxela** [comment]
> Is there a list of items for manual testing somewhere?

Not that I knew. jQuery is used in _so many_ places (all forms, dialogs, ajax, views UI ...).

All we have is the checklist for [jQuery UI](https://github.com/backdrop/backdrop-issues/wiki/jQuery-UI:-Checklist-for-manual-tests), but that's only a subset, of course.

**Comment by indigoxela** [comment]
It will be a challenge for some contrib projects, I found one, that won't be compatible (even with our compat layer).
https://github.com/backdrop-contrib/views_slideshow_cycle2/issues/19

**Comment by indigoxela** [comment]
Looking at contrib... IF we find removed functionality used in contrib, but not used in core - should we add it to the compat layer or not?

Example views_slideshow_cycle2 would additionally need:

```
  $.type = function (obj) {
    return typeof obj;
  }
  $.now = Date.now;
```
Or is it sufficient to inform maintainers in their issue queues?

**Comment by argiepiano** [comment]
While more work for the maintainer and possibly the bug squad, I'd lean towards informing the maintainer. I can see these issues not being fixed for a long time or ever if we did add those to the compat layer.

**Comment by indigoxela** [comment]
Rebased for a fresh sandbox. 😉 It might be, that 3.8 gets released before 4.x stable, we'll see...

**Comment by izmeez** [comment]
Second beta of jQuery 4.0.0 [recently released](https://blog.jquery.com/2024/07/17/second-beta-of-jquery-4-0-0/)

**Comment by indigoxela** [comment]
> Second beta of jQuery 4.0.0 recently released

It's already available in this PR. 😉

**Comment by indigoxela** [comment]
Here we go, updated to jquery-4.0.0-rc.1.min.js, which has been released yesterday.

After applying the patch, go to admin/config/development/jquery and switch to jQuery 4 via admin UI.

To verify the version, open a brower console and drop in `jQuery.fn.jquery`.

To test, click around and do some stuff - jQuery's everywhere. Everything drag-and-drop (layouts, fields...), everything dialog (editor image dialog, file library. views...), everything re dynamic form handling, everything AJAX...

All those things should work and there shouldn't be any console nagging.

**Comment by avpaderno** [comment]
I did some tests on the live preview (except, probably, AJAX). All worked as expected.

**Comment by indigoxela** [comment]
jQuery 4 stable has been released yesterday, so I rebased and updated my PR. 🎉

**Comment by indigoxela** [comment]
Another rebase, hopefully we'll get that into our next minor (1.34). 😉 

Note that jQuery 4 isn't enabled by default, you'd have to switch the version on admin/config/development/jquery

<img width="741" height="401" alt="Image" src="https://github.com/user-attachments/assets/8fd30640-094a-435d-b2f2-9687e832fa8f" />

**Comment by herbdool** [comment]
I've tested with just core: Views, Layouts, etc. It works.

I've suggested renaming `jquery-4-bc.js` to `jquery-4-compat.js` to match the existing pattern for other files.

I also added a question of what we'll do when jQuery 5 comes out. Will it be the new `latest` key and move jQuery 4 to `default`? If so we'll need to make sure we're confident that we won't break anything by forcing sites onto a different version. I don't know what is the right answer here, nor if that changes what we name the new key.

**Comment by indigoxela** [comment]
> I've suggested renaming jquery-4-bc.js to jquery-4-compat.js to match the existing pattern for other files.

Done. 👍 

> I also added a question of what we'll do when jQuery 5 comes out.

That's indeed a good point. As Backdrop 1.x is a very long lived version, we might even see jQuery 5 with it, who knows. 😉
So the term "latest" is a bit too ambiguous. I've now switched to "jquery4" as key.

Also needs testing again because of renaming, I guess.

**Comment by indigoxela** [comment]
Some help with wording would be great.

Here's what we have now:

<img width="773" height="343" alt="Image" src="https://github.com/user-attachments/assets/c8641126-527f-4cbf-872f-e008cea8f6f9" />

Obviously v3 isn't "The latest" anymore. But how should we describe v4 - if we add some description there?

**Comment by olafgrabienski** [comment]
> Some help with wording would be great.

How about something like this?

> [x] jQuery 3 (recommended)
> The default version of jQuery in Backdrop CMS.
>
> [ ] jQuery 1 (legacy)
> Alternative for sites with modules or themes which are not working properly with the recommended version.
>
> [ ] jQuery 4 (experimental)
> The latest version of jQuery.

**Comment by izmeez** [comment]
Is it also possible to have the options listed in some numeric order such as descending: 4, 3, 1 ?

**Comment by herbdool** [comment]
We don't want to change the existing keys. I think we should just focus on the new ones.

Regardless, at some point we'll want to have `default` to v4 or later once we're certain that they're robust. At that point we'd probably want to create a `v3` key so existing sites stay on it. Maybe we could even do that now. V4 would be `default`.

**Comment by indigoxela** [comment]
> Regardless, at some point we'll want to have default to v4 or later once we're certain that they're robust.

Changing the default seems like a 2.x topic to me. We can't just switch without admin interaction in 1.x.
In 2.x we'll probably remove jQuery 1.x and move 3.x or 4.x (or even 5.x) to "default". But we can't do that now.

**Comment by indigoxela** [comment]
@olafgrabienski many thanks for your suggestions. 🙏  I applied them now.

If this PR makes it in, we should probably also update the docs page: https://docs.backdropcms.org/documentation/jquery-in-backdrop-cms.

**Comment by herbdool** [comment]
Actually we already changed the default in 1.x one time so doing it again wouldn't necessarily be 2.x. We would have an update hook to make sure that existing sites stayed on 3.x and 4.x would apply to new sites. 

In this alternative approach we would only do this if we were confident that new sites were fairly solid on 4.x.

**Comment by quicksketch** [comment]
Finally took a look at this. Thank you @indigoxela for the work on this!

I read through the issue and it seems no one has suggested _not_ including the backwards-compatibility file? How would people feel about omitting that and modifying the libraries in core as-needed to make them jQuery 4.x compatible?

Then there would some motivation for contrib modules to get updated, and since jQuery 4.x is not the default (yet), we won't be breaking anyone's site that uses jQuery 3.x.

**Comment by izmeez** [comment]
Are you suggesting dropping jQuery 1.x?

**Comment by quicksketch** [comment]
> Are you suggesting dropping jQuery 1.x?

No, the PR contains a `core/misc/jquery-4-compat.js` file that adds jQuery 3.x functions to jQuery 4.x so code does not need to be updated to be 4.x compatible. I'm suggesting that we not include a compatibility file since 4.x won't be the default. Once contrib is updated to be mostly compatible, we can make 4.x the default.

**Comment by quicksketch** [comment]
A note that [Drupal decided to fork jquery.form.js into core](https://www.drupal.org/project/drupal/issues/3419734), for both security reasons and compatibility with jQuery 4.x.

**Comment by indigoxela** [comment]
> ...backwards-compatibility file? How would people feel about omitting that and modifying the libraries in core as-needed to make them jQuery 4.x compatible?

That's an option, too, of course. We'd take over responsibility for 4 more libraries. 3 of them are currently minified, which raises the question, how to handle that. especially with regards to future modifications. We'd probably need both variants - unminified/minified. And we'd have to document all the steps and in how their handling differs from "regular" 3rd party libs.
(As an example: how to update jquery.ui is largely "internal information". I have a script locally to fetch and build. Not ideal.)

Taking over is a bit more effort, though.

> Then there would some motivation for contrib modules to get updated...

From my experience: "motivation" is (almost) never the cause. It's either that maintainers hesitate to fiddle in JS code they're not familiar with (in module's custom JS). Or - that's more likely - they sit on a library not compatible and not receiving updates anymore.

In the latter case, all of them would have to either take over library maintenance (very unlikely) or provide exactly the backwards-compatibility layer we currently attempt to provide in core. 😉

FWIW - personally, I'm fine with either approach (compat layer vs. "hack").

**Comment by indigoxela** [comment]
Oh, and there's another option for contrib I didn't mention, yet: jquery.migrate.
There's a half done module in contrib to provide such a utility module, which other contrib projects could use as dependency, if they're not able to get their code or lib compatible. Here's an [issue with a suggestion](https://github.com/backdrop-contrib/jquery_migrate/issues/1).

👆  Note: that's about contrib. In core we still have 2 options (compat layer or take over maintenance).

**Comment by indigoxela** [comment]
@quicksketch any thoughts on forking those 3rd party libs and documentation about handling (my previous comments)? We're running out of time for a PR rewrite. 😉

**Comment by quicksketch** [comment]
@indigoxela This may be too ambitious, but here's a possibility.

@jenlampton claimed the https://github.com/jquery-contrib namespace some time ago. We could do the following:

- Fork the needed libraries into https://github.com/jquery-contrib from their current locations on GitHub.
- Update the libraries for jQuery 4.x compatibility in the full source version.
- Compress each library in their GitHub repos.
- Make new releases, incrementing the major version numbers.
- Copy the released versions back into Backdrop core.

Is that too crazy?

@jenlampton Can you make me an owner of that organization so I can assist? I can't add new members at the moment.

**Comment by indigoxela** [comment]
> This may be too ambitious, but here's a possibility.

It's indeed ambitious, given that we're just a few days from feature freeze for 1.34.

Getting appropriate sources for those 4 libs, creating repos, getting code compatible and, not to underestimate, proper testing for this PR here. I'm quite skeptical, this will work for our next release.

Those 3rd party repos (if they still exist), are usually in a structure, that requires a significant amount of time to familiarize with. Quite a learning curve. And we only have the upcoming weekend for that. Who would do that, anyway? Only me?

In the meantime we have an approach that just works and has been tested.

Postponing jQuery 4 support in core until September or even January next year means, people still can't have it with Backdrop CMS. And maintainers still have no chance to check their module's compatibility. Although this PR exists for ... quite a while now. And jQuery 4 stable has been released months ago.

**Comment by quicksketch** [comment]
Agreed @indigoxela, it's probably not feasible. Maybe instead we plan for a
conditional checkbox on the jQuery version page for "Include compatibility
libraries" to toggle the addition of these kinds of BC layers. We already
have similar things for other versions of jQuery and it might be nice to be
able to turn those off. But that option would need to have core be fully
4.x compatible first.

**Comment by indigoxela** [comment]
A checkbox, or possibly different variants (with / without BC) 🤔 
At the moment that won't work, though, as core wouldn't be compatible - jquery.form.js is loaded all over the place. We're running in circles here. 😜

But why is our compat layer such a "problem", anyway? It just extends jQuery in the official, documented way. With 5 unproblematic, simple, straight forward utility functions, 2 of them just aliases to vanilla JavaScript methods... 🤷 

What we *could* do is to rename the lib key to something like "jquery4-bc". So in a future Backdrop version we could add a plain "jquery4" and no modifications would be necessary for settings.

**Comment by herbdool** [comment]
Hopefully we can reach consensus on this issue. Though I don't think it needs to be rushed for 1.34.0; it's just going to be there as an experimental option. If it helps, how about only including `jquery-4.0.0.min.js` but do not expose it in the UI? Then there's time to figure out the rest.

**Comment by herbdool** [comment]
Another thing we'll need consensus on is when to make jquery 3 the legacy version. When we added jquery 3 we added an update hook so existing sites would use the new `legacy` version (jquery 1):

```
/**
 * Set jQuery to the backwards-compatible, legacy version for existing sites.
 */
function system_update_1091() {
  if (config_get('system.core', 'jquery_version') === NULL) {
    config_set('system.core', 'jquery_version', 'legacy');
  }
}
```

I'm partial to only including jquery 4 when we're confident that *new* sites can default to jquery 4. We add an update hook to update existing sites that are on jquery 3 to a new `jquery3` key and make jquery 4 the new `default` key. This would make it consistent with our previous approach. (This decision is independent of whether to add a compat layer or fork the unmaintained libraries).

**Comment by herbdool** [comment]
For self-edification I looked up the old jquery 3 issue https://github.com/backdrop/backdrop-issues/issues/3106 after a dev meeting discussion. We had a question of why we chose `default` and `legacy` and seems like I had just added it to the PR and people agreed.

**Comment by quicksketch** [comment]
We discussed this at the weekly meeting today.

My opinion on this: I think we should move forward with the PR with the compatibility layer. Keep jQuery 3.x as the default for now.

We probably should change the version keys. Instead of `default` and `legacy`, we should switch to `1.x`, `3.x`, and `4.x`. That should have an update hook to change the config keys.

**Comment by indigoxela** [comment]
> Hopefully we can reach consensus on this issue. Though I don't think it needs to be rushed for 1.34.0;

@herbdool I'm pretty sure, we have consensus, it's only about implementation details here. "Rush" is an odd word for a PR that exists for over 2 years now. 😉

I see no way to make jQuery 4 the default any time soon. Way too many contrib projects (or custom code) could be affected. But if we wait until each and every contrib is compatible, we'll probably wait forever.
And in the meantime it would be tricky for contrib maintainers (or site builders) to even figure out, whether a project is compatible or not. For exactly this case I encourage to get this into core soon, so people can start to evaluate (and act, if necessary). The benefit of making it in steps (first provide v4, not enabled by default, later make it the default) should be obvious. At least, I hope it is.

Re the keys... yeah. "default" and "legacy" have been chosen at a time, when it didn't look like there would be movement on the jQuery project. But there is for a while now. jQuery's alive! 😉 

So, the keys should get switched to (literal):

- 1.x
- 3.x
- 4.x

Question: should we append some indication for the bc layer to that "4.x" or not?

**Comment by indigoxela** [comment]
Ouch, possible problem: changing the existing version keys might have impact on hook_library_info_alter() implementations.

Or any contrib/custom check for `$config->get('jquery_version')` 😬 

A quick (sloppy) check: two projects in contrib would be affected - navbar (module) and standford_decanter (install profile).

**Comment by indigoxela** [comment]
@herbdool and @quicksketch could we make the lib key change and update hook a follow-up? It feels unrelated here (has nothing to do with v4), and has some potential for a breaking change, that needs proper consideration and discussion.

I switched to `4.x` for the newly added, but wouldn't want to introduce a breaking change re existing keys.

**Comment by herbdool** [comment]
@indigoxela yes, the key change can be a follow-up. Can you create the new issue to capture a summary of what to do?

**Comment by quicksketch** [comment]
Great work @indigoxela! Thank you for your tremendous patience. Thank you @herbdool for the final review and suggestions!

I've merged https://github.com/backdrop/backdrop/pull/4657 into 1.x .

Thank you @klonos, @avpaderno, @izmeez, @argiepiano, and @olafgrabienski for the help in feedback and testing!

**Comment by indigoxela** [comment]
Follow-up for key change of existing versions has been created here: https://github.com/backdrop/backdrop-issues/issues/7115
